src/mam4xx/gas_chem.hpp

```diff
@@ -0,0 +1,391 @@
+    // ! ... the non-linear component
+    // !-----------------------------------------------------------------------
+
+    if (factor[nr_iter]) {
+      nlnmat(sys_jac, //   & ! out
+             lin_jac,
+             dti); // ! in
+      // -----------------------------------------------------------------------
+      //  ... factor the "system" matrix
+      // -----------------------------------------------------------------------
+
+      lu_fac(sys_jac);
+
+    } // factor
+    // !-----------------------------------------------------------------------
+    // ! ... form f(y)
+    // !-----------------------------------------------------------------------
+    imp_prod_loss(prod, loss,        // & ! out
+                  lsol, lrxt, lhet); // ! in
+
+    for (int mm = 0; mm < clscnt4; ++mm) {
+      forcing[mm] =
+          solution[mm] * dti - (iter_invariant[mm] + prod[mm] - loss[mm]);
+    } // mm
+
+    // !-----------------------------------------------------------------------
+    // ! ... solve for the mixing ratio at t(n+1)
+    // !-----------------------------------------------------------------------
+    lu_slv(sys_jac, forcing);
+    for (int mm = 0; mm < clscnt4; ++mm) {
+      solution[mm] += forcing[mm];
+    } // mm
+
+    // !-----------------------------------------------------------------------
+    // ! ... convergence measures
+    // !-----------------------------------------------------------------------
+
+    if (nr_iter > 0) {
+      for (int kk = 0; kk < clscnt4; ++kk) {
+        int mm = permute_4[kk];
+        // printf("mm %d \n", mm);
+        // BAD CONSTANT
+        if (haero::abs(solution[mm]) > 1.e-20) {
+          max_delta[kk] = haero::abs(forcing[mm] / solution[mm]);
+        } else {
+          max_delta[kk] = zero;
+        }
+
+      } // kk
+
+    } // nr_iter
+
+    // !-----------------------------------------------------------------------
+    // ! ... limit iterate
+    // !-----------------------------------------------------------------------
+    for (int kk = 0; kk < clscnt4; ++kk) {
+      if (solution[kk] < zero) {
+        solution[kk] = zero;
+      }
+    } // end kk
+
+    // !-----------------------------------------------------------------------
+    // ! ... transfer latest solution back to work array
+    // !-----------------------------------------------------------------------
+
+    for (int kk = 0; kk < clscnt4; ++kk) {
+      int jj = clsmap_4[kk];
+      int mm = permute_4[kk];
+      lsol[jj] = solution[mm];
+    } // end kk
+
+    // !-----------------------------------------------------------------------
+    // ! ... check for convergence
+    // !-----------------------------------------------------------------------
+
+    if (nr_iter > 0) {
+      convergence = true;
+      for (int kk = 0; kk < clscnt4; ++kk) {
+        converged[kk] = true;
+
+        int mm = permute_4[kk];
+        bool frc_mask = haero::abs(forcing[mm]) > small ? true : false;
```

Comment:
Would it be cleaner to leave the ternary operator out of this?
`bool frc_mask = haero::abs(forcing[mm]) > small`